LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClientExtensions.cs

```diff
@@ -0,0 +1,125 @@
+                                                       CancellationToken cancellationToken,
+                                                       [CallerMemberName] string? operationName = null) =>
+                InvokeAsync(operation, Void.Value, Void.Value, cancellationToken, operationName);
+
+            private Task<TResult> InvokeAsync<T, TResult>(Operation<T, Void, TResult> operation, T arg,
+                                                          CancellationToken cancellationToken,
+                                                          [CallerMemberName] string? operationName = null) =>
+                InvokeAsync(operation, arg, Void.Value, cancellationToken, operationName);
+
+            private async Task<TResult> InvokeAsync<T1, T2, TResult>(Operation<T1, T2, TResult> operation, T1 arg1, T2 arg2,
+                                                                     CancellationToken cancellationToken,
+                                                                     [CallerMemberName] string? operationName = null)
+            {
+                for (var attempt = 1; ; attempt++)
+                {
+                    try
+                    {
+                        _ = this.client.EnsureConnected();
+                        return await operation(this.client, arg1, arg2, cancellationToken);
+                    }
+                    catch (Exception ex)
+                        when ((ex is ObjectDisposedException
+                                  || (ex is InvalidOperationException ioe
+                                      && ioe.Message.StartsWith("This operation is only allowed using a successfully authenticated context.", StringComparison.OrdinalIgnoreCase)))
+                              && ExceptionFilterUtility.True(() =>
+                                     this.logger?.LogWarning(ex, @"Device client operation ""{Operation}"" (attempt {Attempt}/"
+                                                                 + MaxAttempts.ToString(CultureInfo.InvariantCulture)
+                                                                 + @") failed due to error: {Error}",
+                                                                 operationName, attempt, ex.GetBaseException().Message)))
+                    {
+                        // disconnect, re-connect and then retry...
+                        await this.client.DisconnectAsync(CancellationToken.None);
+                        if (attempt == MaxAttempts)
+                            throw;
+                    }
+                }
+            }
+
+            private delegate Task<TResult> Operation<in T1, in T2, TResult>(ILoRaDeviceClient client, T1 arg1, T2 arg2, CancellationToken cancellationToken);
+
+            private static class Operations
+            {
+                public static readonly Operation<Void, Void, Twin> GetTwin = (client, _, _, cancellationToken) => client.GetTwinAsync(cancellationToken);
```

Comment:
We had a chat on this one: I felt it was a lot of added complexity to avoid the additional allocations on the closure, but runtime wise, it is a good option. Atif and I agreed, using a static function to avoid capturing any context and avoid additional allocations.

Reply:
I agree and so with 43b1b882938e30d1eaf06be1572f501505cbd98a, I (logically) reverted to the initial and simpler implementation I had in 37575aa475152e55b172d379cd20b15954ea317f.